### PR TITLE
congure: initial import of a congestion control framework

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This provides _CongURE_ (*Cong*estion Control *U*tilizing *R*eusable *E*lements, pronounced like "conjure"), a framework to implement congestion control and use it in a reusable manner with a number of protocols.

The original motivation was to provide congestion control with [6LoWPAN selective fragment recovery](https://doc.riot-os.org/group__net__gnrc__sixlowpan__frag__sfr.html) and allow for the congestion control mechanism to be interchangeable. However, now that there is a framework with multiple well tested implementations of congestion control, it can also be used for other protocols, such as TCP or QUIC, as I kept the units for the congestion window up to the user (TCP and QUIC, e.g., uses bytes, SFR, e.g., uses number of fragments). It is also designed as such that multiple instances of the same implementation can run on the same image (e.g. for when TCP and SFR want to use the same CC). In that case, I can imagine a lot of ROM can be saved, since there is only one CC code for two layers.

A test framework for the CongURE framework is also provided in form of shell commands to trigger the API functions of CongURE, as well as tests for said framework (a framework test framework test if you want :wink:). If preferred, I can move those to a separate PR.

As to why this is in `sys/` and not `sys/net/`: There are no networking components to this framework, it just provides the functions to calculate the size of a congestion window. `seq` e.g. is also in `sys/`, though sequence number arithmetic mostly shows up in networking. Not sure this could help with local queue congestion problems, but when thinking out of the box, one might think as such :-P.

### Testing procedure
A test framework including tests for it is provided. [`riotctrl`](https://github.com/RIOT-OS/riotctrl/) (at least v0.2.2) is required for them to run. If installed,

```
make -j -C tests/congure_test flash test
```

should succeed.

### Issues/PRs references
Follow-up PRs for actual implementations of congestion control mechanisms will follow, as well as integration PRs for GNRC's 6LoWPAN SFR and maybe TCP as well :-). When merged, I will adapt the test framework to use the turo framework introduced in #15950, which was the original idea to begin with (which is why it is already printing JSON strings ;-)).
PR dependency graph

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
Is there any whitelist/black list needed for the boards or features required? A quick test on the `nucleo-f411re` has timeouts (even with just plain term).

<details>

```
BOARD=nucleo-f411re make term -C tests/congure_test/
make: Entering directory '/home/weiss/wd/RIOT/tests/congure_test'
/home/weiss/wd/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-02-09 10:33:17,797 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
help
2021-02-09 10:33:19,835 # Command              Description
2021-02-09 10:33:19,838 # ---------------------------------------
2021-02-09 10:33:19,844 # state                Prints current CongURE state object as JSON
2021-02-09 10:33:19,847 # reboot               Reboot the node
2021-02-09 10:33:19,852 # version              Prints current RIOT_VERSION
2021-02-09 10:33:19,857 # pm                   interact with layered PM subsystem
2021-02-09 10:33:19,861 # cong_clear           Clears CongURE state object
2021-02-09 10:33:19,867 # cong_setup           Calls the setup function for the CongURE state object
2021-02-09 10:33:19,873 # cong_init            Calls init method of the CongURE state object
2021-02-09 10:33:19,881 # cong_imi             Calls inter_message_interval method of the CongURE state object
2021-02-09 10:33:19,887 # cong_report          Calls a report_* method of the CongURE state object
> state
2021-02-09 10:33:23,914 #  {"driver":"0x00000000","ctx":"0x00000000","cwnd":0,"init":{"calls":0,"last_args":{"c":"0x00000000","ctx":"0x00000000"}},"inter_msg_interval":{"calls":0,"last_args":{"c":"0x00000000","msg_size":0}},"report_msg_sent":{"calls":0,"last_args":{"c":"0x00000000","msg_size":0}},"report_msg_discarded":{"calls":0,"last_args":{"c":"0x00000000","msg_size":0}},"report_msg_timeout":{"calls":0,"last_args":{"c":"0x00000000","msgs":[]}},"report_msg_lost":{"calls":0,"last_args":{"c":"0x00000000","msgs":[]}},"report_msg_acked":{"calls":0,"last_args":{"c":"0x00000000","msg":null,"ack":null}},"report_ecn_ce":{"calls":0,"last_args":{"c":"0x00000000","time":0,}},}
pm
2021-02-09 10:33:29,816 #  Usage:
2021-02-09 10:33:29,821 # 	pm show: display current blockers for each power mode
2021-02-09 10:33:29,827 # 	pm set <mode>: manually set power mode (lasts until WFI returns)
2021-02-09 10:33:29,831 # 	pm block <mode>: manually block power mode
2021-02-09 10:33:29,835 # 	pm unblock <mode>: manually unblock power mode
2021-02-09 10:33:29,837 # 	pm off: call pm_off()
> version
2021-02-09 10:33:31,524 #  2021.04-devel-474-g90b0f-congure/feat/initial
> cong_clear
2021-02-09 10:34:02,223 #  > main(): This is RIOT! (Version: 2021.04-devel-474-g90b0f-congure/feat/initial)
help
2021-02-09 10:34:07,452 #  Command              Description
2021-02-09 10:34:07,455 # ---------------------------------------
2021-02-09 10:34:07,462 # state                Prints current CongURE state object as JSON
2021-02-09 10:34:07,464 # reboot               Reboot the node
2021-02-09 10:34:07,469 # version              Prints current RIOT_VERSION
2021-02-09 10:34:07,474 # pm                   interact with layered PM subsystem
2021-02-09 10:34:07,478 # cong_clear           Clears CongURE state object
2021-02-09 10:34:07,484 # cong_setup           Calls the setup function for the CongURE state object
2021-02-09 10:34:07,490 # cong_init            Calls init method of the CongURE state object
2021-02-09 10:34:07,498 # cong_imi             Calls inter_message_interval method of the CongURE state object
2021-02-09 10:34:07,504 # cong_report          Calls a report_* method of the CongURE state object
> cong_init
2021-02-09 10:34:16,036 #  {"error":"State object not set up"}
> cong_setup
2021-02-09 10:34:35,819 #  {"setup":"0x20000284"}
> cong_init
2021-02-09 10:34:39,189 #  {"error":"`ctx` argument expected"}
> cong_init a
2021-02-09 10:34:42,053 #  {"error":"`ctx` expected to be hex"}
> cong_init 0
2021-02-09 10:35:36,993 # Exiting Pyterm
```
</details>

**Comment by miri64** [comment]
> Is there any whitelist/black list needed for the boards or features required?

Shouldn't be, which is why so far, I only tested on `native`. Will try to reproduce on one of the platforms I have at hand.

**Comment by miri64** [comment]
Where do I see the timeout in your output?

**Comment by miri64** [comment]
(`cong_init` does not print anything on success, if that is what you mean...)

**Comment by MrKevinWeiss** [comment]
It doesn't show a prompt and locks or something compared to native.

**Comment by miri64** [comment]
> It doesn't show a prompt and locks or something compared to native.

Mh... two problems, I did not check the length of the argument, before accessing the array, which I fixed. However, it seems like the prompt is not flushed when a command has no output in general:

```
2021-02-09 12:53:02,690 #  main(): This is RIOT! (Version: 2021.04-devel-474-g90b0f-congure/feat/initial)
> cong_setup
2021-02-09 12:53:06,004 #  {"setup":"0x20000284"}
> cong_init 0x0
help
2021-02-09 12:53:18,596 #  > shell: command not found: he
> cong_setup
2021-02-09 12:53:31,069 #  {"setup":"0x20000284"}
> cong_init 0x0
help
2021-02-09 12:53:38,535 #  > Command              Description
2021-02-09 12:53:38,539 # ---------------------------------------
2021-02-09 12:53:38,545 # state                Prints current CongURE state object as JSON
2021-02-09 12:53:38,548 # reboot               Reboot the node
2021-02-09 12:53:38,552 # version              Prints current RIOT_VERSION
2021-02-09 12:53:38,557 # pm                   interact with layered PM subsystem
2021-02-09 12:53:38,562 # cong_clear           Clears CongURE state object
2021-02-09 12:53:38,568 # cong_setup           Calls the setup function for the CongURE state object
2021-02-09 12:53:38,574 # cong_init            Calls init method of the CongURE state object
2021-02-09 12:53:38,582 # cong_imi             Calls inter_message_interval method of the CongURE state object
2021-02-09 12:53:38,588 # cong_report          Calls a report_* method of the CongURE state object
2021-02-09 12:54:13,481 # Exiting Pyterm
```

Note the `>` prompt showing up in the table head of the `help` command. Will dig a bit why.

**Comment by miri64** [comment]
Seems to be an issue with `pyterm`. When using `RIOT_TERMINAL=socat`, this is not a problem:

```
cong_setup
{"setup":"0x20000284"}
> cong_init 0x0
> help
Command              Description
---------------------------------------
state                Prints current CongURE state object as JSON
reboot               Reboot the node
version              Prints current RIOT_VERSION
pm                   interact with layered PM subsystem
cong_clear           Clears CongURE state object
cong_setup           Calls the setup function for the CongURE state object
cong_init            Calls init method of the CongURE state object
cong_imi             Calls inter_message_interval method of the CongURE state object
cong_report          Calls a report_* method of the CongURE state object
```

Will have a look if I find a fix.

**Comment by miri64** [comment]
Rebased to include https://github.com/RIOT-OS/RIOT/pull/15962.

**Comment by miri64** [comment]
Great... now when hitting enter on an empty command, does not work anymore...

**Comment by miri64** [comment]
Ah, RIOTCtrl needs to use `cleanterm` instead of `term` for this to work :-).

**Comment by miri64** [comment]
AAaand that's not configurable with RIOTCtrl :rofl:

**Comment by miri64** [comment]
See https://github.com/RIOT-OS/riotctrl/pull/16

**Comment by miri64** [comment]
Ok, but besides of that, I guess I won't be using `pyterm` for these tests... 

With the new DEBUG feature enabled, I get the following with `pyterm`:

```
------------------------------------------------------------------ Captured log call ------------------------------------------------------------------
DEBUG    TestCongUREWithoutSetup:01-run.py:44 '\x1b[?2004l\r\x1b[?2004h{"error":"State object not set up"}'
```

So same problem as https://github.com/RIOT-OS/RIOT/issues/15888...

**Comment by miri64** [comment]
With the last commit I am able to run it on `iotlab-m3` (even without RIOT-OS/riotctrl#16)

**Comment by miri64** [comment]
(but merging RIOT-OS/riotctrl#16 would not hurt either way)

**Comment by miri64** [comment]
Rebased and squashed to resolve conflict in `dist/tools/uncrustify/whitelist.txt`.

**Comment by MrKevinWeiss** [comment]
It works now with the `nucleo-f411re`

**Comment by miri64** [comment]
Changed some naming (and corresponding doc) I was not happy with.

**Comment by miri64** [comment]
And optimized some of the structs for size.

**Comment by miri64** [comment]
image

**Comment by miri64** [comment]
I leave it for tomorrow to port all the tests to EOL Python 3.5 compatible code...

**Comment by miri64** [comment]
But first, I moved the CongURE mock, I introduced in `tests/congure_test` to its own module `congure_mock`, as I found myself copying it to test the usage in SFR.

**Comment by miri64** [comment]
And replaced all the f-Strings :-/

**Comment by MrKevinWeiss** [comment]
> And replaced all the f-Strings :-/

Always a sad thing to do.

**Comment by miri64** [comment]
@kaspar030 can we install the newest `riotctrl` on Murdock's Pi fleet, please? Why aren't the tests there run from our docker?

**Comment by miri64** [comment]
Thanks, @kaspar030 for installing `riotctrl` and `pytest` on the pifleet!

**Comment by miri64** [comment]
Mhhh ESP32 is doing weird things again in Murdock :confused: 

<details>
```
=================================== FAILURES ===================================
__________________ TestCongUREWithoutSetup.test_init_wo_setup __________________

self = <01-run.TestCongUREWithoutSetup testMethod=test_init_wo_setup>

    def test_init_wo_setup(self):
        res = self.exec_cmd('cong_init 0x12345')
>       self.assertEqual(res, {'error': "State object not set up"})
E       AssertionError: None != {'error': 'State object not set up'}

tests/01-run.py:141: AssertionError
----------------------------- Captured stdout call -----------------------------
Makefile:28: warning: overriding recipe for target 'test'
/tmp/dwq.0.6046216988587385/31a939da415d345f232bc1da918d1174/makefiles/tests/tests.inc.mk:22: warning: ignoring old recipe for target 'test'
socat - open:/dev/ttyUSB0,b115200,echo=0,raw 
ets Jun  8 2016 00:22:57



rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)

configsip: 0, SPIWP:0xee

clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00

mode:DOUT, clock div:2

load:0x3fff0018,len:4

load:0x3fff001c,len:4004

load:0x40078000,len:7304

load:0x40080000,len:7224

entry 0x40080350



main(): This is RIOT! (Version: buildtest)
> cong_clear
> cong_init 0x12345
> 
------------------------------ Captured log call -------------------------------
DEBUG    TestCongUREWithoutSetup:01-run.py:47 ''
___________ TestCongUREWithoutSetup.test_inter_msg_interval_wo_setup ___________

self = <01-run.TestCongUREWithoutSetup testMethod=test_inter_msg_interval_wo_setup>

    def test_inter_msg_interval_wo_setup(self):
        res = self.exec_cmd('cong_imi 689')
>       self.assertEqual(res, {'error': "State object not set up"})
E       AssertionError: None != {'error': 'State object not set up'}

tests/01-run.py:145: AssertionError
----------------------------- Captured stdout call -----------------------------
cong_clear
{"error":"State object not set up"}
> cong_imi 689
> 
------------------------------ Captured log call -------------------------------
DEBUG    TestCongUREWithoutSetup:01-run.py:47 ''
```

</details>

The output shown for `cong_clean` in `TestCongUREWithoutSetup.test_inter_msg_interval_wo_setup` is actually the output that is expected for `cong_init 0x12345` by `TestCongUREWithoutSetup.test_init_wo_setup`. Seems like the output is off by one command :-/

**Comment by miri64** [comment]
(I can't reproduce this locally, btw, but I have not tried with ESP32 connected to a Raspberry Pi yet.

**Comment by miri64** [comment]
> (I can't reproduce this locally, btw, but I have not tried with ESP32 connected to a Raspberry Pi yet.

Did that now, but can't reproduce either:

```

Operating System Environment
----------------------------
         Operating System: "Raspbian GNU/Linux" "10 (buster)"
                   Kernel: Linux 4.9.80-v7+ armv7l unknown
             System shell: /bin/dash (probably dash)
             make's shell: /bin/dash (probably dash)

Installed compiler toolchains
-----------------------------
               native gcc: gcc (Raspbian 8.3.0-6+rpi1) 8.3.0
        arm-none-eabi-gcc: missing
                  avr-gcc: missing
         mips-mti-elf-gcc: missing
           msp430-elf-gcc: missing
       riscv-none-elf-gcc: missing
  riscv64-unknown-elf-gcc: missing
     riscv-none-embed-gcc: missing
     xtensa-esp32-elf-gcc: missing
   xtensa-esp8266-elf-gcc: missing
                    clang: missing

Installed compiler libs
-----------------------
     arm-none-eabi-newlib: missing
      mips-mti-elf-newlib: missing
        msp430-elf-newlib: missing
    riscv-none-elf-newlib: missing
riscv64-unknown-elf-newlib: missing
  riscv-none-embed-newlib: missing
  xtensa-esp32-elf-newlib: missing
xtensa-esp8266-elf-newlib: missing
                 avr-libc: missing (missing)

Installed development tools
---------------------------
                   ccache: missing
                    cmake: missing
                 cppcheck: missing
                  doxygen: missing
                      git: git version 2.20.1
                     make: GNU Make 4.2.1
                  openocd: missing
                   python: Python 2.7.16
                  python2: Python 2.7.16
                  python3: Python 3.7.3
                   flake8: error: /usr/bin/python3: No module named flake8
               coccinelle: missing
```

**Comment by miri64** [comment]
Ok, I blacklisted the `esp32-wroom-32` for testing for now and will open an issue with a detailed bug description later (or a PR with a reproducibility test case).

**Comment by miri64** [comment]
> Ok, I blacklisted the `esp32-wroom-32` for testing for now and will open an issue with a detailed bug description later (or a PR with a reproducibility test case).

https://github.com/RIOT-OS/RIOT/pull/16014

**Comment by fjmolinas** [comment]
Bumping https://github.com/RIOT-OS/RIOT/pull/16026, sorry.

**Comment by miri64** [comment]
No problem, I bumped it after #16026 already before. Just want to have this at least run once today (it does not use the unittests, and I just want to investigate a potential bug in `riotctrl` with the CI ;-))

**Comment by kaspar030** [comment]
hmm, 82k builds?

**Comment by kaspar030** [comment]
> (it does not use the unittests, and I just want to investigate a potential bug in `riotctrl` with the CI ;-))

ok :)

**Comment by miri64** [comment]
Ok, increasing the timeout to wait for the initial prompt seems to work. This needs to be fixed in `riotctrl`.

**Comment by miri64** [comment]
> hmm, 82k builds?

Well in theory I only need to rerun the tests, but currently I don't have any other way to reproduce this and the code that is not changed since the last build is rebuild regardless ;-).

**Comment by kaspar030** [comment]
> Well in theory I only need to rerun the tests, but currently I don't have any other way to reproduce this and the code that is not changed since the last build is rebuild regardless ;-).

btw, you can limit the build run by hard-editing BOARDS and/or APPS in .murdock. preferably in a commit that makes CI checks fail. ;)

**Comment by miri64** [comment]
> Ok, increasing the timeout to wait for the initial prompt seems to work. This needs to be fixed in `riotctrl`.

See https://github.com/RIOT-OS/riotctrl/pull/19

**Comment by miri64** [comment]
> btw, you can limit the build run by hard-editing BOARDS and/or APPS in .murdock.

Thanks for the hint.

> preferably in a commit that makes CI checks fail. ;)

Well in this case I was aiming for success :P

**Comment by miri64** [comment]
Let's check if 0.5 sec, the new default in riotctrl 0.4, are already enough for the ESP32 or if we should go to 1sec again

**Comment by miri64** [comment]
Reviewers, after that: may I rebase?

**Comment by miri64** [comment]
Retoggling, just to make sure the .5 timing is enough ;-)

**Comment by MrKevinWeiss** [comment]
Yes you may squash

**Comment by miri64** [comment]
Ok, remaining issue should be fixed once https://github.com/RIOT-OS/riotdocker/pull/135 is merged. I could wait for #16037 to make the `congure_test` application fully Kconfig compliant but I would double push for that for better traceability anyways. So will squash now and later rebase when #16037 is merged and adopt the test then :-).

**Comment by miri64** [comment]
But first some Kconfig fixes to bring it in line with the style established in #16037.

**Comment by miri64** [comment]
And squashed (and removed some of the test commits)

**Comment by miri64** [comment]
`mobi6` needs a new docker :-/

**Comment by miri64** [comment]
Now it works! :woman_dancing:

**Comment by miri64** [comment]
Removed that test commit.

**Comment by miri64** [comment]
And rebased to current master to include #16037.

**Comment by miri64** [comment]
> And rebased to current master to include #16037.

And I can't use it, since using just Kconfig and Kconfig + dependency resolution via Kconfig for an application (so having an `app.config` and an `app.config.test`) does not work at the moment :-/.

**Review by MrKevinWeiss — COMMENTED** [review]
Generally I would say things look pretty beautiful. There is a lot here but you probably will be using that. Tests are provided, there is enough documentation, I am sure things will be more understandable as additional implementations get added but it is nice to start with the mock.

Tests pass with native but I am having some trouble with the board tests, I will investigate.

Just to clarify, some error values are -1 because they are used in general coding and other error values are 1 because they will be used as a shell command (sys/include/congure/test.h)?

**Comment by MrKevinWeiss** [comment]
The board tests seem to pass now...

<details><summary> BOARD=nucleo-f411re make clean flash test -C tests/congure_test/ </summary>

```
BOARD=nucleo-f411re make clean flash test -C tests/congure_test/
make: Entering directory '/home/weiss/wd/RIOT/tests/congure_test'
Building application "tests_congure_test" for "nucleo-f411re" with MCU "stm32".

"make" -C /home/weiss/wd/RIOT/boards/nucleo-f411re
"make" -C /home/weiss/wd/RIOT/boards/common/nucleo
"make" -C /home/weiss/wd/RIOT/core
"make" -C /home/weiss/wd/RIOT/cpu/stm32
"make" -C /home/weiss/wd/RIOT/cpu/cortexm_common
"make" -C /home/weiss/wd/RIOT/cpu/cortexm_common/periph
"make" -C /home/weiss/wd/RIOT/cpu/stm32/periph
"make" -C /home/weiss/wd/RIOT/cpu/stm32/stmclk
"make" -C /home/weiss/wd/RIOT/cpu/stm32/vectors
"make" -C /home/weiss/wd/RIOT/drivers
"make" -C /home/weiss/wd/RIOT/drivers/periph_common
"make" -C /home/weiss/wd/RIOT/sys
"make" -C /home/weiss/wd/RIOT/sys/auto_init
"make" -C /home/weiss/wd/RIOT/sys/congure
"make" -C /home/weiss/wd/RIOT/sys/congure/mock
"make" -C /home/weiss/wd/RIOT/sys/congure/test
"make" -C /home/weiss/wd/RIOT/sys/fmt
"make" -C /home/weiss/wd/RIOT/sys/isrpipe
"make" -C /home/weiss/wd/RIOT/sys/malloc_thread_safe
"make" -C /home/weiss/wd/RIOT/sys/newlib_syscalls_default
"make" -C /home/weiss/wd/RIOT/sys/pm_layered
"make" -C /home/weiss/wd/RIOT/sys/shell
"make" -C /home/weiss/wd/RIOT/sys/shell/commands
"make" -C /home/weiss/wd/RIOT/sys/stdio_uart
"make" -C /home/weiss/wd/RIOT/sys/test_utils/interactive_sync
"make" -C /home/weiss/wd/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  22204	    132	   2540	  24876	   612c	/home/weiss/wd/RIOT/tests/congure_test/bin/nucleo-f411re/tests_congure_test.elf
/home/weiss/wd/RIOT/dist/tools/openocd/openocd.sh flash /home/weiss/wd/RIOT/tests/congure_test/bin/nucleo-f411re/tests_congure_test.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: using old version of OpenOCD
Using STLINK_VERSION 2-1
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
adapter speed: 2000 kHz
adapter_nsrst_delay: 100
none separate
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
srst_only separate srst_nogate srst_open_drain connect_assert_srst
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : clock speed 1800 kHz
Info : STLINK v2 JTAG v30 API v2 SWIM v19 VID 0x0483 PID 0x374B
Info : using stlink api v2
Info : Target voltage: 3.252984
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       reset
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
adapter speed: 1800 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x080012f4 msp: 0x20000200
auto erase enabled
Info : device id = 0x10006431
Info : flash size = 512kbytes
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x20000046 msp: 0x20000200
wrote 32768 bytes from file /home/weiss/wd/RIOT/tests/congure_test/bin/nucleo-f411re/tests_congure_test.elf in 1.275677s (25.085 KiB/s)
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000002e msp: 0x20000200
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000002e msp: 0x20000200
verified 22336 bytes in 0.232695s (93.739 KiB/s)
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
adapter speed: 1800 kHz
shutdown command invoked
Done flashing
......................................
----------------------------------------------------------------------
Ran 38 tests in 16.333s

OK
make: Leaving directory '/home/weiss/wd/RIOT/tests/congure_test'
````

</details>

**Comment by miri64** [comment]
> Just to clarify, some error values are -1 because they are used in general coding and other error values are 1 because they will be used as a shell command (sys/include/congure/test.h)?

Yes.

File: sys/include/congure/test.h [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+ *
+ * Every byte in the state object is set to 0.
+ *
+ * @param[in] argc  Number of @p argv. Needs to be at least 1.
+ * @param[in] argv  Command line arguments. No extra arguments are required
+ *                  except for the command name in `argv[0]`.
+ *
+ * Provides no output.
+ *
+ * @return  Always 0.
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/include/congure/test.h:107]: Because this is a shell command handler. Those need an `int` return value.

**Comment by MrKevinWeiss** [comment]
How can I run this with pytest for a board?

File: sys/include/congure/test.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (miri64) [reply sys/include/congure/test.h:107]: The "Provides no output." part is not true anymore though. Will fix ASAP.

**Comment by MrKevinWeiss** [comment]
It seems like `pytest tests/congure_test/tests/01-run.py` just times out...

**Comment by miri64** [comment]
> It seems like `pytest tests/congure_test/tests/01-run.py` just times out...

Is `BOARD` set in your environment?

**Comment by MrKevinWeiss** [comment]
With and without all fail.

**Comment by MrKevinWeiss** [comment]
exporting as well.  Do I need to have some `--add-opts` or something for the `BOARD`?

**Comment by miri64** [comment]
> exporting as well. Do I need to have some `--add-opts` or something for the `BOARD`?

No, let me try.

File: sys/include/congure/test.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (miri64) [reply sys/include/congure/test.h:108]: Fixed.

**Comment by miri64** [comment]
With the latest fixes it should now also work on `samr21-xpro` again. Seems like having too long commands breaks something...

**Comment by miri64** [comment]
> It seems like `pytest tests/congure_test/tests/01-run.py` just times out...

Ah! You have to execute it within the test directory. So

```sh
cd tests/congure_test
pytest tests/01-run.py
```

Should I add a fix for that to be possible from any directory, or should we leave it as is, since this is not integrated anyways (yet, but when it will be, it will be used via `make` anyways where we have `CURDIR`)

**Comment by MrKevinWeiss** [comment]
> Should I add a fix for that to be possible from any directory, or should we leave it as is, since this is not integrated anyways (yet, but when it will be, it will be used via make anyways where we have CURDIR)

Maybe just add a note on the readme.

**Comment by MrKevinWeiss** [comment]
image

**Comment by miri64** [comment]
> > Should I add a fix for that to be possible from any directory, or should we leave it as is, since this is not integrated anyways (yet, but when it will be, it will be used via make anyways where we have CURDIR)
> 
> Maybe just add a note on the readme.

Done

**Review by MrKevinWeiss — CHANGES_REQUESTED** [review]
Tests work with make and pytest on both native and nucleo-f411re.

It seems like there are some kconfig issues:

<details>

```
make clean
weiss@mobiweiss:~/wd/RIOT/tests/congure_test$ TEST_KCONFIG=1 make all
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-=> The parameter KCONFIG_USEMODULE_SHELL could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: USEMODULE_SHELL (=n), (!TEST_KCONFIG) (=n)

[genconfig.py]:ERROR-=> The parameter SHELL_NO_ECHO could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: ((KCONFIG_USEMODULE_SHELL && !TEST_KCONFIG) || (MODULE_SHELL && TEST_KCONFIG)) (=n)

make: *** No rule to make target '/home/weiss/wd/RIOT/tests/congure_test/bin/native/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
```
</details>

Maybe @leandrolanzieri can give some feedback.

**Comment by MrKevinWeiss** [comment]
same thing if I say `make menuconfig`

**Comment by miri64** [comment]
> It seems like there are some kconfig issues:

Mhhh I thought I solved that already :-/ Will have another look, but I think it's because with `TEST_KCONFIG=1` it still tries to load the config for `app.config`.

**Comment by miri64** [comment]
@leandrolanzieri would it make sense that with `TEST_KCONFIG=1` we only consider `*.config.test` files, while with `!TEST_KCONFIG` we only consider `*.config` files in the application?

**Comment by miri64** [comment]
> > It seems like there are some kconfig issues:
> 
> Mhhh I thought I solved that already :-/ Will have another look, but I think it's because with `TEST_KCONFIG=1` it still tries to load the config for `app.config`.

Yepp, if I remove the `app.config` file in the test, it works:

```console
$ TEST_KCONFIG=1 make -C tests/congure_test/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test'
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-=> The parameter KCONFIG_USEMODULE_SHELL could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: USEMODULE_SHELL (=n), (!TEST_KCONFIG) (=n)

make: *** No rule to make target '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test/bin/native/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
make: *** Waiting for unfinished jobs....
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test'
$ rm tests/congure_test/app.config
$ TEST_KCONFIG=1 make -C tests/congure_test/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test'
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_congure_test" for "native" with MCU "native".

"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/core
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/native/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/auto_init
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/congure
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/fmt
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native/periph
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers/periph_common
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native/stdio_native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/test_utils/interactive_sync
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/congure/mock
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/congure/test
/usr/bin/ld: /home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test/bin/native/cpu/tramp.o: warning: relocation against `_native_saved_eip' in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
   text	   data	    bss	    dec	    hex	filename
  47915	    772	  48004	  96691	  179b3	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test/bin/native/tests_congure_test.elf
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/congure_test'
```

**Comment by MrKevinWeiss** [comment]
Just got there.. yup, it looks like the `app.config` is being loaded when `TEST_KCONFIG=1`

**Comment by MrKevinWeiss** [comment]
Could it have something to do with #16064?

**Comment by MrKevinWeiss** [comment]
probably not...

**Comment by miri64** [comment]
> Could it have something to do with #16064?

No, I don't think so. This is purely about `app.config` being included all the time:

https://github.com/RIOT-OS/RIOT/blob/f3871c06e7ca04b1807f8150f781d38dd43b2565/makefiles/kconfig.mk#L58-L65

**Comment by miri64** [comment]
Oh, I did not even push the `app.config.test` yet, I just noticed Oo

**Comment by MrKevinWeiss** [comment]
I wonder if we just drop trying to support the old way and only keep the post migration way...

I guess it depends on if app.config would be able to be used with app.config.test (like it is now) or if it should be either or.

**Comment by miri64** [comment]
> I wonder if we just drop trying to support the old way and only keep the post migration way...
> 
> I guess it depends on if app.config would be able to be used with app.config.test (like it is now) or if it should be either or.

This will make full migration a pain, I believe, since we then have to port all the `app.config`s in one big effort. It would be better if they could exist side by side for a while IMHO. See https://github.com/RIOT-OS/RIOT/pull/16092

**Comment by MrKevinWeiss** [comment]
Please rebase and I will retest with the app.config.test fix

**Comment by miri64** [comment]
> Please rebase and I will retest with the app.config.test fix

Rebased and squashed. Remember the clean step inbetween though ;-)

**Comment by miri64** [comment]
Rebased and squashed

**Review by MrKevinWeiss — APPROVED** [review]
I think things look good to me.  Kconfig is checked, the idea is documented, test work for a board and native.  ACK!

**Comment by miri64** [comment]
Thanks for the review and all the help with Kconfig migration and RIOTCtrl deployment!

**Pull request merged**